sphinx_immaterial/custom_admonitions.py

```diff
@@ -0,0 +1,377 @@
+        textnodes, messages = self.state.inline_text(title_text, self.lineno)
+        if "no-title" not in self.options and title_text:
+            title = nodes.title(title_text, "", *textnodes)
+            title.source, title.line = self.state_machine.get_source_and_line(
+                self.lineno
+            )
+            admonition_node += title
+        admonition_node += messages
+        admonition_node["classes"] += self.classes
+        self.state.nested_parse(self.content, self.content_offset, admonition_node)
+        return [admonition_node]
+
+
+def get_directive_class(name, title, classes=None) -> Type[CustomAdmonitionDirective]:
+    """A helper function to produce a admonition directive's class."""
+    # alias upstream-deprecated CSS classes for pre-defined admonitions in sphinx
+    class_list = [nodes.make_id(name)]
+    if classes:
+        class_list.extend(classes)
+
+    # uncomment this block when we merge v9.x from upstream
+    # if name in ("caution", "attention"):
+    #     class_list.append("warning")
+    # elif name == "error":
+    #     class_list.append("danger")
+    # elif name in ("important", "hint"):
+    #     class_list.append("tip")
+    # elif name == "todo":
+    #     class_list.append("info")
+
+    class CustomizedAdmonition(CustomAdmonitionDirective):
+        default_title = title
+        classes = class_list
+        optional_arguments = int(name not in admonitionlabels)
+        node_class = nodes.admonition if name != "todo" else sphinx.ext.todo.todo_node
+
+    return CustomizedAdmonition
+
+
+def on_builder_inited(app: Sphinx):
+    """register the directives for the custom admonitions and build the CSS."""
+    config = app.config
+    custom_admonitions: List[CustomAdmonitionConfig] = getattr(
+        config, "sphinx_immaterial_custom_admonitions"
+    )
+    setattr(app.builder.env, "sphinx_immaterial_custom_icons", {})
+    for admonition in custom_admonitions:
+
+        app.add_directive(
+            name=admonition.name,
+            cls=get_directive_class(
+                admonition.name, admonition.title, admonition.classes
+            ),
+            override=admonition.override,
+        )
+
+        # set variables for CSS template to match HTML output from generated directives
+        admonition.name = nodes.make_id(admonition.name)
+        if admonition.icon is not None:
+            admonition.icon = load_svg_into_builder_env(app.builder, admonition.icon)
+    setattr(app.builder.env, "sphinx_immaterial_custom_admonitions", custom_admonitions)
+
+
+def on_config_inited(app: Sphinx, config: Config):
+    """Add admonitions based on CSS classes inherited from mkdocs-material theme."""
+
+    # override the generic admonition directive
+    app.add_directive("admonition", get_directive_class("admonition", ""), True)
+
+    # generate directives for inherited admonitions from upstream CSS
+    if getattr(config, "sphinx_immaterial_generate_inherited_admonitions"):
```

Comment:
Perhaps instead of "inherited" this should be "extra"?

Reply:
This makes sense from a user perspective.